web-extensions/chrome/engine/bookmycomics.js

```diff
@@ -37,66 +56,105 @@ BmcEngine.prototype.events = {
+ * whether the comicId was already resolved or not.
+ *
+ */
+BmcEngine.prototype._memoizeComic = function () {
+    if (!this._comic.name) {
+	console.warn('BmcEngine._memoizeComic: unknown comic name.');
+        return;
+    }
+    if (this._comic.id !== undefined) {
+	// console.log(`BmcEngine._memoizeComic: Cache Hit (${this._comic.id}).`);
+        this._dispatchLoad();
+	return ;
+    }
+    if (this._comic.memoizing) {
+	// console.log('BmcEngine._memoizeComic: Memoize request already pending.');
+	return ;
+    }
+    this._comic.memoizing = true;
+    this._db.findComic(this._comic.name, (err, comicId) => {
+	// console.log('BmcEngine._memoizeComic: memoized id=', comicId);
+        this._comic.id = comicId;
+	this._comic.memoizing = false;
+        this._dispatchLoad();
+    });
+}
+
+/*
+ * This function is the main window's entry point to setup the add-on's
+ * necessary utilities (UI, background process, etc.)
+ * It requires the current page's comis info, for any UI that might need to be
+ * spawned and would require it.
+ *
+ */
+BmcEngine.prototype.setup = function() {
+    return this._ui.makeSidePanel(
+        () => this.track(),
+        this._comic.name, this._comic.chapter, this._comic.page);
+}
+
 /*
  * This function is the entrypoint for a reader page, which spawns the right UI
  * piece according to the situation.
  *
  * The reader's page must send any available information relative to the
  * displayed comic (if any).
  *
- * @params {String} comicName - Name of the comic according to the reader
- * @params {Number} chapter - Current chapter of the comic according to the
- *                            reader's page
- * @params {Number|null} page - Current page of the comic according to the
- *                              reader's page
- *
  */
-BmcEngine.prototype.track = function(comicName, chapter, page) {
-    console.log(`BookMyComic: bmcEngine.track: manga=${comicName} chapter=${chapter} page=${page}`);
-    this._db.findComic(comicName, (err, comicId) => {
-        if (err) {
-            console.log('BookMyComic: Could not track comic: Database error (step1)');
-            return ;
-        }
-        console.log(`BookMyComic: bmcEngine.track: Got comicId from storage: ${comicId}`);
-        this._ui.makeSidePanel(comicName, chapter, page);
-        if (comicId === null) {
-            return this._ui.makeRegisterDialog(comicName, chapter, page);
+BmcEngine.prototype.track = function() {
+    console.log(`BookMyComic: bmcEngine.track: manga=${this._comic.name} chapter=${this._comic.chapter} page=${this._comic.page}`);
+
+    // One single way to handle this, whether the id was already memoized or
+    // not: use the event handling.
+    console.warn(`Attempting track: event=${this.events.load}`);
+    this.addEventListener(this.events.load, () => {
+        console.log(`BookMyComic: bmcEngine.track.doTrack: Got comicId from storage: ${this._comic.id}`);
+        if (this._comic.id === null) {
```

Comment:
I'm a bit worry here. The initialization of this value is [here](https://github.com/Joacchim/BookMyComics/pull/2/files#diff-74c1ba5acd49c483b50977afad619d48R28) and is `undefined`. Shouldn't you test `undefined` as well?

Reply:
The point is that with the memoization, the EventListener will only be triggered when the value is resolved. In which case, the "not found" value is set to null.
Undefined is the init value used to differenciate unresolved and not found.